### ISSUE TITLE
OS hardlock when getting search results

### ISSUE THREAD

**Issue by sylvainfaivre** [nicotine-plus/nicotine-plus#2128]
Nicotine+ version:  initially 3.2.2, same with latest daily
Operating System/Distribution: Ubuntu 20.04.4 LTS

Nicotine+ was previously running fine.
Since yesterday, when doing a search, the whole OS freezes after a few seconds. This is a hardlock, I can't even switch to a tty, I have to power cycle.

I tried to enable the debug log, but nothing gets written to it before the hardlock.
I also tried different keywords for search, but it fails anyway, as soon as the results are a few pages.
If I search for a very specific track title that gives only a few results, this bug does not happen.

Any hints about how to find the cause of this bug ?

BTW, I checked CPU & RAM usage, CPU temperature, even checked the RAM with MEMtest86+. Everything is normal.

**Comment by slook** [comment]
Please specify your GTK version (Menu > Help > About Nicotine+)?

It will be useful check your OS update logs to see if any of your installed packages that are Gtk related were recently upgraded, if so then it might be a bug in Gtk that we need to work around, but that is only a guess.

> Any hints about how to find the cause of this bug ?

You might have to help with testing, because it will be difficult for us to replicate the bug unless it is related to something that affects all Ubuntu LTS users.

**Comment by sylvainfaivre** [comment]
> Please specify your GTK version (Menu > Help > About Nicotine+)?

This is GTK 3.24.20

> It will be useful check your OS update logs to see if any of your installed packages that are Gtk related (or any of it's dependencies) were recently upgraded, if so then it might be a bug in Gtk that we need to work around, but that is only a guess.

Well, I had not run Nicotine+ for 3 months before this session, so there were some gtk related packages update, for example xdg-desktop-portal-gtk:amd64, gir1.2-javascriptcoregtk-4.0:amd64, libwebkit2gtk-4.0-37:amd64, libjavascriptcoregtk-4.0-18:amd64, libwebkit2gtk-4.0-37:amd64, libjavascriptcoregtk-4.0-18:amd64...

What makes it weirder is, when I started this particular session, eveything went well for about 30 minutes, then I started getting this bug, and I've always got it since then.

This makes me wonder if there is something triggering this problem in one of my local files. I tried deleting the `~/.local/share/nicotine` folder, but this changed nothing.

I also noticed I had the `[searches] max_displayed_results` set to 5000, I thought maybe this was too much, so I tried 1000 and even tried 100, but I still get this bug.

> If it were a crash inside Nicotine+ itself then there would be a Critical Error dialog with a traceback, but this doesn't seem to be the case so it will be challenging to find clues.

OK well I'm going to upgrade to Ubuntu 22.04.1 LTS as soon as it is released, which should happen in a few days. We'll see if it fixes this bug.

> You could [run from source](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md#git) then that gives you stdout to the console for any potential `print("test")` or `log.add_debug("test")` statements that could be added into the code to help us pinpoint where things go wrong.

OK thanks, I'll try this ASAP.

> You might have to help with testing, because it will be difficult for us to replicate the bug unless it is related to something that affects all Ubuntu LTS users, which seems unlikely but not impossible.

**Comment by sylvainfaivre** [comment]
> > You could [run from source](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md#git) then that gives you stdout to the console for any potential `print("test")` or `log.add_debug("test")` statements that could be added into the code to help us pinpoint where things go wrong.
> 
> > You might have to help with testing, because it will be difficult for us to replicate the bug unless it is related to something that affects all Ubuntu LTS users, which seems unlikely but not impossible.
>
> OK thanks, I'll try this ASAP.

Ooops sorry, I didn't read that through. I'm willing to help if you think of stuff to try, and I can run from source with added debug statements if needed. Otherwise I'll wait a few days for the Ubuntu 22.04.1 LTS and report back after upgrading.

**Comment by sylvainfaivre** [comment]
PS. Ubuntu 22.04.1 LTS is scheduled for release on August 4.

**Comment by slook** [comment]
Ideally it makes sense for the 3.2.x branch to be stable even though your distro version is soon going to be obsolete, because it is possible that other Debian variants that have slower update and release cycles could also become affected by this bug at some point, and due to the severity of the crash it is the sort of thing that would make a new user abandon the app.

> If I search for a very specific track title that gives only a few results, this bug does not happen.

I have a feeling it might be something trivial related to GtkTreeView scrolling that triggers some bug in an external library that doesn't belong to us. If that is the case then this should be reported to upstream or to the Ubuntu LTS team, because this sort of thing that brings the host machine down should never happen on any system no matter how old it is. Everybody would agree that it would be a shame for the final release of the old stable Ubuntu branch (or the old stable Nicotine+ 3.2.x branch) to have some crippling bug in it that bricks the installation for legacy users.

> I had the `[searches] max_displayed_results` ... and even tried 100, but I still get this bug.

Good idea, please count the exact number of visible rows on page of the results list that avoids rendering the vertical scroll bar, and then go 1 more in order to try to establish if the crash is triggered by TreeView scrolling or not.

**Comment by mathiascode** [comment]
> I thought maybe this was too much, so I tried 1000 and even tried 100, but I still get this bug.

In that case, I doubt this is related to the GUI. My guess is that the number of connections performed to other peers when searching for popular terms is causing issues on your system.

If you're able to run SoulseekQt, could you confirm if the issue occurs there as well?

**Comment by sylvainfaivre** [comment]
> > You could [run from source](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md#git) then that gives you stdout to the console...
> > > OK thanks, I'll try this ASAP.
> 
> If you run Nicotine+ this way then Gtk Warning error messages might be displayed in the console window just before the systems hangs, so have the console window visible during runtime testing to potentially enable the capturing any debugging information in a photograph.

OK I just ran it from the git clone, I get the same problem, and no log displayed before crashing (besides the usual logs).

> > I can't even switch to a tty,
> 
> Do you have a way of testing if the machine is still up? If it still accessible over a network then the display manager is what crashed. On Debian systems you can usually restart the X11 window manager session by pressing Ctrl+Alt+Backspace, but on Ubuntu system this feature needs to be enabled first, see Ubuntu forum: [How can I enable Ctrl+Alt+Backspace?](https://askubuntu.com/questions/10622/how-can-i-enable-ctrlaltbackspace)

When this happens, the machine is hardlocked, it doesn't answer to ping / ssh any more, or to Alt-F* (change tty).

**Comment by sylvainfaivre** [comment]
> In that case, I doubt this is related to the GUI. To completely rule out the GUI, close Nicotine+, and set `max_displayed_results` in `~/.config/nicotine/config` to 0.
> 
> My guess is that the number of connections performed to other peers when searching for popular terms is causing issues on your system.
> 
> If you're able to run SoulseekQt, could you confirm if the issue occurs there as well?

Well, you're right, the bug is not related to the GUI. I set `max_displayed_results` to 0 and the bug still happens.
I also tried starting over with an empty config file, and I get the same result.

I just tried with Soulseek AppImage from https://www.slsknet.org/news/node/1 and it works without any problem for the same search (I run my tests with the "korn freak" query, which gives thousands of results).

This is getting weirder by the minute...

What's really strange is I started experiencing this bug after using Nicotine+ for 30 minutes, seraching and downloading files without any problem !

**Comment by mathiascode** [comment]
> especially when Connection Debug [Conn] logging to file is enabled!

22c44fd9a79b731ebbda042d66a309cfa52ce7c9 should improve this a lot.

Could you test some older Nicotine+ releases and see if they all display the same symptoms? The easiest way would be to download "Source code (tar.gz)" for a release and run `./nicotine`: https://github.com/nicotine-plus/nicotine-plus/releases

Please back up your ~./config/nicotine and ~./local/share/nicotine folders first though, as downgrading can screw up some things when you switch back to a recent version.

**Comment by sylvainfaivre** [comment]
Thank you both for your help.
I didn't find the debug log options before, but thanks to you, I got some logs now !
They don't seem to show anything special for me, but maybe they will for you.

Here's a bit of explanation to help with the logs : 
The station running Nicotine is hooked to a TV screen, and I pilot it through VNC.
I reproduced the bug 2 times and took screncaps in VNC when the machine locked, these are bad1.png and bad2.png.
I also took pictures of the TV screen with my phone, because some logs were written to the screen and never reached the VNC client. These are bad1s.png and bad2s.png.
As you will see, bad1.png ends at the middle of bad1s.png, while bad2s.png has nearly a full screen of logs after bad2.png ends.

I also added good.png, which is the logs I get with a smaller search (that doesn't crash Nicotine+).

I'm sending these by email right now.

**Comment by sylvainfaivre** [comment]
> Could you test some older Nicotine+ releases and see if they all display the same symptoms? 

Sure, I'm stopping here for now, but I'll try older releases in the next days.

**Comment by mathiascode** [comment]
Unfortunately the logs don't display any obvious pattern. Does the connection counter in the status bar always freeze on a high number, and if so, how many connections are there?

**Comment by jsuttinger** [comment]
This started happening for me yesterday. I can here looking for this issue and was happy to see it reported.  As soon as I search in Nicotine+ my entire OS locks up and I have to power cycle. Debug logs are also not written before I power cycle.

However, I am able to reproduce the issue with SoulSeekQT (SSQT). The same behavior is displayed; open SSQT, immediately search for anything, entire OS is froze and must be power cycled.

OS: Linux Mint 20.3
Nicotine+: 3.2.2
SSQT: 2018-1-30

Since it's reproduced in both applications, probably a recent update that was pushed to our OS's or something. I haven't had time to dig into it too much.

**Comment by mathiascode** [comment]
The issue seems to be exclusive to Ubuntu 20.04 and derivatives so far.

I assume the system is fully up-to-date? I know there was a critical kernel bug in Ubuntu 20.04 earlier this month, but I'm not sure if it's related.

**Comment by jsuttinger** [comment]
> The issue seems to be exclusive to Ubuntu 20.04 and derivatives so far.
> 
> I assume the system is fully up-to-date? I know there was a critical kernel bug in Ubuntu 20.04 earlier this month, but I'm not sure if it's related.

Besides the MySQL database client library update that popped up today, yes it's up to date. 
I'm running Kernel: Linux 5.4.0-122-generic

**Comment by mathiascode** [comment]
I found this bug report, which was created a day before the Nicotine+ one: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1982014

Could you try downgrading the kernel to 5.4.0-121 and see if that's the culprit?

**Comment by jsuttinger** [comment]
Downgraded to 5.4.0-121
Issue does not persist. I'll keep an eye on it and report back if it does. However the issue was so easy to reproduce I suspect it was related to kernel 5.4.0-122

**Comment by mathiascode** [comment]
A new kernel version with a fix will be released on August 1 according to https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1981658

**Comment by jsuttinger** [comment]
Awesome, thanks for the insight.

**Comment by sylvainfaivre** [comment]
Hi, thanks for the pointer to the kernel bug !
I was running 5.4.0-122-generic, too, and installing 5.4.0-123-generic from the "proposed" repo fixed my problem.
So this was a kernel bug, not a Nicotine+ bug :-)

**Comment by slook** [comment]
Even though this bug does not belong to Nicotine+ we will leave this issue open until 5.4.0-123 in released into Ubuntu stable and is confirmed fixed.

**Comment by slook** [comment]
Info about this kernel bug: 

[[PATCH] tcp: md5: incorrect tcp_header_len for incoming connections](https://lkml.kernel.org/netdev/CANn89iLL38HMB2EtaSc9Xt436xg9Jke-3LT+KCt9B6ktxj1yCw@mail.gmail.com/T/#t)

[[PATCH net] tcp: make sure treq->af_specific is initialized](https://lore.kernel.org/all/20220424203509.2801158-1-eric.dumazet@gmail.com/#r)

**Comment by tatiee** [comment]
Can someone do an ELI5 on how to do either of these:

    downgrade the kernel to linux-image-5.4.0-121-generic
    upgrade the kernel to linux-image-5.4.0-123-generic from the "proposed" repo

Went to DL some albums today and happened to run into this same issue. I'm fairly new to Linux unfortunately...

**Comment by slook** [comment]
Sorry we cannot provide support here for things that are beyond the scope of Nicotine+.

You could try looking around on [Ask Ubuntu](https://askubuntu.com/questions/1404722/downgrade-kernel-for-ubuntu-22-04-lts) or on the [Ubuntu Forums](https://ubuntuforums.org/showthread.php?t=2475637) for the answer to your question @tatiee , fortunately the Linux community is helpful but nothing is guaranteed... Iirc last time I used `apt` to install a custom kernel something broke!

If your system boots with GRUB and has multiple kernels installed (for example left over from previous updates etc) then you might choose which one to load with the [Advanced Options](https://askubuntu.com/questions/1014634/how-to-access-advanced-options-in-grub) menu during startup.

> I'm fairly new to Linux unfortunately...

Package management is not really much fun for a beginner, and modifying your primary installation with an alternative kernel might not be wise unless you know exactly what you are doing and have a spare machine/installation incase something goes wrong...

It could be easier just to make a Live USB/DVD to run/install a different distro onto a spare drive/partition such as [MX Linux](https://mxlinux.org/) (which btw has a selection of tried and tested kernels to choose from in their stable repo without using the command line). Distro hopping is a great way to learn about Linux, whereas breaking your system's package management is not.

**Comment by mathiascode** [comment]
`linux-image-5.4.0-124-generic` containing a fix should now be available.
